doc/man/man9/time.9.man
.so man9/osfcompat.9.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) doc/man/man9/time.9.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database strcompat.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so strcompat.macros
.\"
.\"
.TH TIME 9 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ DDI/DKI"
.\"
.\"
.SH NAME
.B time
\- time in seconds since epoch
.\"
.\"
.SH SYNOPSIS
.PP
.B #define _OSF_SOURCE
.br
.B #include <sys/ddi.h>
.br
.B #include <sys/osfddi.h>
.HP
int \fBtime\fR(int \fIyyy\fR);
.\"
.\"
.SH ARGUMENTS
.PP
.TP
.I yyy
.\"
.\"
.SH DESCRIPTION
.PP
.BR time ()
.\"
.\"
.SH RETURN
.PP
When
.BR time ()
succeeds, it returns 
.BR DDI_SUCCESS ;
when it fails, it returns
.BR DDI_FAILURE .
.\"
.\"
.SH ERRORS
.PP
.BR time ()
fails when 
.\"
.\"
.SH CONTEXT
.PP
.BR time ()
can be called from
.\"
.\"
.SH MP-STREAMS
.PP
.\"
.\"
.SH NOTICES
.PP
.BR time ()
is a
.IR OSF/1 \(rg-specific
function:
.B _OSF_SOURCE
must be defined before including DDI/DKI include files to expose this symbol.
.B @PACKAGE_TITLE@
must also have been compiled with
.B CONFIG_STREAMS_COMPAT_OSF
or
.B CONFIG_STREAMS_COMPAT_OSF_MODULE
defined.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR STREAMS (9).
.\"
.\"
.SH BUGS
.PP
.BR time ()
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
.BR time ()
is compatible with
.IR OSF/1 \(rg
with the following portability considerations:
.IP \(em \w'\(em\(em'u
.IP \(em
.IP \(em
.IP \(em
.BR time ()
is provided for source level compatibility with
.IR OSF/1 \(rg
for the purpose of porting
.IR OSF/1 \(rg
drivers to
.BR Linux .
Because internal structures and definitions for
.B @PACKAGE_TITLE@
.[
@PACKAGE@-@VERSION@
.]
differ from that of
.IR OSF/1 \(rg,
binaries are incompatible.
.IP \(em
.BR time ()
only provides functionality comparable to
.IR OSF/1 \(rg
for use by
.I STREAMS
drivers and modules.
Any and all functionality provided by
.BR time ()
in support of
.RI non- STREAMS
devices are not supported.
.PP
See
.BR STREAMS (9)
for additional compatibility considerations.
.\"
.\"
.SH CONFORMANCE
.PP
.IR OSF/1 \(rg
DDI documentation.
.[
osfdoc
.]
.[
osfspg
.]
.\"
.\"
.SH HISTORY
.PP
.BR time ()
first appeared in
.IR OSF/1 \(rg.
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


isode/dsap/common/time.c
/*****************************************************************************

 @(#) $RCSfile$ $Name$($Revision$) $Date$

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2007  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date$ by $Author$

 -----------------------------------------------------------------------------

 $Log$
 *****************************************************************************/

#ident "@(#) $RCSfile$ $Name$($Revision$) $Date$"

static char const ident[] = "$RCSfile$ $Name$($Revision$) $Date$";

/* time.c - */

#ifndef	lint
static char *rcsid =
    "Header: /xtel/isode/isode/dsap/common/RCS/time.c,v 9.0 1992/06/16 12:12:39 isode Rel";
#endif

/* 
 * Header: /xtel/isode/isode/dsap/common/RCS/time.c,v 9.0 1992/06/16 12:12:39 isode Rel
 *
 *
 * Log: time.c,v
 * Revision 9.0  1992/06/16  12:12:39  isode
 * Release 8.0
 *
 */

/*
 *				  NOTICE
 *
 *    Acquisition, use, and distribution of this module and related
 *    materials are subject to the restrictions of a license agreement.
 *    Consult the Preface in the User's Manual for the full terms of
 *    this agreement.
 *
 */

/*
	SYNTAX:
		time ::= 'yymmddhhmmssz'

		where yy   = year
		      mm   = month
		      dd   = day
		      hh   = hours
		      mm   = minutes
		      ss   = seconds
	              z    = timezone

	EXAMPLE:
		890602093221Z -> 09:32:21 GMT, on June 2nd 1989.
*/

#include "quipu/util.h"
#include "quipu/attr.h"
#include "psap.h"

extern int strprint();
extern int sfree();
extern int pstrcmp();

static UTC
qstr2utct(s, len)
	char *s;
	int len;
{
	UTC ut;

	if (len > 14 && strncmp(s, "1989", 4) == 0 && (ut = str2utct(s + 2, len - 2)))
		return ut;

	return str2utct(s, len);
}

#define	str2utct	qstr2utct

static PE
timeenc(x)
	char *x;
{
	PE ret_pe = NULLPE;

	/* Should switch to pepsy -> need to use qbufs! */

	(void) build_UNIV_UTCTime(&ret_pe, 0, 0, x, NULL);
	return (ret_pe);
}

static char *
timedec(pe)
	PE pe;
{
	char *x;

	if (parse_UNIV_UTCTime(pe, 0, 0, &x, NULL) == NOTOK)
		return (NULLCP);
	return (x);
}

utcprint(ps, xtime, format)
	PS ps;
	char *xtime;
	int format;
{
	UTC ut;

	if (format == READOUT && (ut = str2utct(xtime, strlen(xtime)))) {
		long mtime;

		mtime = gtime(ut2tm(ut));

		ps_printf(ps, "%-24.24s", ctime(&mtime));
	} else
		ps_printf(ps, "%s", xtime);
}

static
utccmp(a, b)
	char *a, *b;
{
	long a_time, mdiff;
	UTC ut;

	if ((ut = str2utct(a, strlen(a))) == NULL)
		return pstrcmp(a, b);
	a_time = gtime(ut2tm(ut));

	if ((ut = str2utct(b, strlen(b))) == NULL)
		return pstrcmp(a, b);

	return ((mdiff = a_time - gtime(ut2tm(ut))) == 0L ? 0 : mdiff > 0L ? 1 : -1);
}

time_syntax()
{
	(void) add_attribute_syntax("UTCTime",
				    (IFP) timeenc, (IFP) timedec,
				    (IFP) strdup, utcprint,
				    (IFP) strdup, utccmp, sfree, NULLCP, NULLIFP, FALSE);
}
